src/audio/midi/MidiAudio.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "MidiAudio.hpp"
+#include "../terminal/parser/stateMachine.hpp"
+
+namespace
+{
+    class MidiOut
+    {
+    public:
+        static constexpr auto NOTE_OFF = 0x80;
+        static constexpr auto NOTE_ON = 0x90;
+        static constexpr auto PROGRAM_CHANGE = 0xC0;
+
+        // We're using a square wave as an approximation of the sound that the
+        // original VT525 terminals might have produced. This is probably not
+        // quite right, but it works reasonably well.
+        static constexpr auto SQUARE_WAVE_SYNTH = 80;
+
+        MidiOut() noexcept
+        {
+            midiOutOpen(&handle, MIDI_MAPPER, NULL, NULL, CALLBACK_NULL);
+            OutputMessage(PROGRAM_CHANGE, SQUARE_WAVE_SYNTH);
+        }
+        ~MidiOut() noexcept
+        {
+            midiOutClose(handle);
+        }
+        void OutputMessage(const int b1, const int b2, const int b3 = 0, const int b4 = 0) noexcept
+        {
+            midiOutShortMsg(handle, MAKELONG(MAKEWORD(b1, b2), MAKEWORD(b3, b4)));
+        }
+
+        MidiOut(const MidiOut&) = delete;
+        MidiOut(MidiOut&&) = delete;
+        MidiOut& operator=(const MidiOut&) = delete;
+        MidiOut& operator=(MidiOut&&) = delete;
+
+    private:
+        HMIDIOUT handle = nullptr;
+    };
+}
+
+using namespace std::chrono_literals;
+
+MidiAudio::~MidiAudio() noexcept
+{
+    try
+    {
+#pragma warning(suppress : 26447)
+        // We acquire the lock here so the class isn't destroyed while in use.
+        // If this throws, we'll catch it, so the C26447 warning is bogus.
+        _inUseMutex.lock();

```

Comment:
Forgive me for jumping in like this, but isn't using an object while it is being destructed undefined behavior anyway? So this needs to be prevented in some other way?

Reply:
I wouldn't have thought so. Obviously the destructor itself is going to be using the object while it is being destructed, so I don't see why it should be any different for other methods. It's just not something that can be relied on in general, because a destructor may be busy cleaning up resources that the other methods depend on. But that's not the case here.

I may be wrong though. I'm not an expert on the standard. Do you have a reference to somewhere that documents this as undefined behavior? If necessary we could always call the `Lock` method manually before the destructor is reached - that's just a bit messier.

Reply:
<https://docs.microsoft.com/en-us/cpp/standard-library/mutex-class-stl?view=msvc-170#dtormutex_destructor> and <https://en.cppreference.com/w/cpp/thread/mutex/%7Emutex> say that the behavior is undefined if an std::mutex is destroyed while owned. It looks like that will happen here, as MidiAudio::~MidiAudio locks _inUseMutex and leaves it locked and the members are then destroyed.

Reply:
Oh sorry, I misunderstood what @PetterS was saying. If the issue is just that the mutex shouldn't be destroyed while locked, we can just unlock it here as well. We're only acquiring the lock to make sure the playing thread isn't still holding it - we don't need to retain the lock.

Reply:
Alright, thanks. Perhaps it is OK as long as the mutex is also unlocked here.